// xstate/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var actions = require('./actions.js');
var Actor = require('./Actor.js');
var interpreter = require('./interpreter.js');
var Machine = require('./Machine.js');
var mapState = require('./mapState.js');
var match = require('./match.js');
var schema = require('./schema.js');
var State = require('./State.js');
var StateNode = require('./StateNode.js');
var behaviors = require('./behaviors.js');
var types = require('./types.js');
var utils = require('./utils.js');

var assign = actions.assign,
    cancel = actions.cancel,
    send = actions.send,
    sendTo = actions.sendTo,
    sendParent = actions.sendParent,
    sendUpdate = actions.sendUpdate,
    forwardTo = actions.forwardTo,
    doneInvoke = actions.doneInvoke,
    raise = actions.raise,
    log = actions.log,
    pure = actions.pure,
    choose = actions.choose,
    stop = actions.stop;

exports.actions = actions;
exports.toActorRef = Actor.toActorRef;
exports.Interpreter = interpreter.Interpreter;
Object.defineProperty(exports, 'InterpreterStatus', {
  enumerable: true,
  get: function () { return interpreter.InterpreterStatus; }
});
exports.interpret = interpreter.interpret;
exports.spawn = interpreter.spawn;
exports.Machine = Machine.Machine;
exports.createMachine = Machine.createMachine;
exports.mapState = mapState.mapState;
exports.matchState = match.matchState;
exports.createSchema = schema.createSchema;
exports.t = schema.t;
exports.State = State.State;
exports.StateNode = StateNode.StateNode;
exports.spawnBehavior = behaviors.spawnBehavior;
Object.defineProperty(exports, 'ActionTypes', {
  enumerable: true,
  get: function () { return types.ActionTypes; }
});
Object.defineProperty(exports, 'SpecialTargets', {
  enumerable: true,
  get: function () { return types.SpecialTargets; }
});
exports.matchesState = utils.matchesState;
exports.toEventObject = utils.toEventObject;
exports.toObserver = utils.toObserver;
exports.toSCXMLEvent = utils.toSCXMLEvent;
exports.assign = assign;
exports.cancel = cancel;
exports.choose = choose;
exports.doneInvoke = doneInvoke;
exports.forwardTo = forwardTo;
exports.log = log;
exports.pure = pure;
exports.raise = raise;
exports.send = send;
exports.sendParent = sendParent;
exports.sendTo = sendTo;
exports.sendUpdate = sendUpdate;
exports.stop = stop;
