// @rgba-image/common/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var clamp_1 = require("./clamp");
exports.clampByte = clamp_1.clampByte;
exports.clampUint32 = clamp_1.clampUint32;
var consts_1 = require("./consts");
exports.COMPOSITE_NORMAL = consts_1.COMPOSITE_NORMAL;
exports.COMPOSITE_MULTIPLY = consts_1.COMPOSITE_MULTIPLY;
exports.COMPOSITE_SCREEN = consts_1.COMPOSITE_SCREEN;
exports.COMPOSITE_OVERLAY = consts_1.COMPOSITE_OVERLAY;
exports.COMPOSITE_DARKEN = consts_1.COMPOSITE_DARKEN;
exports.COMPOSITE_LIGHTEN = consts_1.COMPOSITE_LIGHTEN;
exports.COMPOSITE_HARD_LIGHT = consts_1.COMPOSITE_HARD_LIGHT;
exports.COMPOSITE_DIFFERENCE = consts_1.COMPOSITE_DIFFERENCE;
exports.COMPOSITE_EXCLUSION = consts_1.COMPOSITE_EXCLUSION;
exports.compositeModeNames = consts_1.compositeModeNames;
exports.compositeModeNameToMode = consts_1.compositeModeNameToMode;
var get_index_1 = require("./get-index");
exports.getIndex = get_index_1.getIndex;
var is_little_endian_1 = require("./is-little-endian");
exports.isLittleEndian = is_little_endian_1.isLittleEndian;
var predicates_1 = require("./predicates");
exports.isImageData = predicates_1.isImageData;
exports.isGrayData = predicates_1.isGrayData;
exports.isRgbaStop = predicates_1.isRgbaStop;
exports.isRgbaUint32Stop = predicates_1.isRgbaUint32Stop;
exports.isChannelStop = predicates_1.isChannelStop;
var uint32_1 = require("./uint32");
exports.rgbaToUint32 = uint32_1.rgbaToUint32;
exports.uint32ToRgba = uint32_1.uint32ToRgba;
